internal/controller/bucket/conversions.go

```diff
@@ -0,0 +1,156 @@
+				for k, v := range l.Properties.AdditionalProperties {
+					o.Labels[i].Properties.AdditionalProperties[k] = v
+				}
+			}
+		}
+		sort.SliceStable(o.Labels, func(i, j int) bool {
+			return o.Labels[i].ID < o.Labels[j].ID
+		})
+	}
+	return o
+}
+
+// GenerateBucket returns a Bucket model that the InfluxDB API accepts for creation
+// and update.
+func GenerateBucket(name string, params v1alpha1.BucketParameters) *domain.Bucket {
+	sType := domain.SchemaType(params.SchemaType)
+	out := &domain.Bucket{
+		Name:        name,
+		Description: params.Description,
+		OrgID:       params.OrgID,
+		Rp:          params.RP,
+		SchemaType:  &sType,
+	}
+	if len(params.RetentionRules) != 0 {
+		out.RetentionRules = make([]domain.RetentionRule, len(params.RetentionRules))
+		for i, rr := range params.RetentionRules {
+			out.RetentionRules[i] = domain.RetentionRule{
+				EverySeconds:              rr.EverySeconds,
+				ShardGroupDurationSeconds: rr.ShardGroupDurationSeconds,
+				Type:                      domain.RetentionRuleType(rr.Type),
+			}
+		}
+	}
+	return out
+}
+
+// LateInitialize sets the defaults from the API if user didn't set a value for
+// such fields.
+func LateInitialize(params *v1alpha1.BucketParameters, obs *domain.Bucket) bool {
+	li := resource.NewLateInitializer()
+	params.Description = li.LateInitializeStringPtr(params.Description, obs.Description)
+	params.RP = li.LateInitializeStringPtr(params.RP, obs.Rp)
+	if params.SchemaType == "" && obs.SchemaType != nil && string(*obs.SchemaType) != "" {
+		params.SchemaType = string(*obs.SchemaType)
+		return true
+	}
+	return li.IsChanged()
+}
+
+// IsUpToDate returns whether an update call is necessary.
+func IsUpToDate(params v1alpha1.BucketParameters, obs *domain.Bucket) bool {
+	if len(params.RetentionRules) != len(obs.RetentionRules) {
+		return false
+	}
+	sort.SliceStable(params.RetentionRules, func(i, j int) bool {
+		return retentionRuleKey(params.RetentionRules[i].Type, params.RetentionRules[i].EverySeconds, params.RetentionRules[i].ShardGroupDurationSeconds) <
+			retentionRuleKey(params.RetentionRules[j].Type, params.RetentionRules[j].EverySeconds, params.RetentionRules[j].ShardGroupDurationSeconds)
+	})
+	sort.SliceStable(obs.RetentionRules, func(i, j int) bool {
+		return retentionRuleKey(string(obs.RetentionRules[i].Type), obs.RetentionRules[i].EverySeconds, obs.RetentionRules[i].ShardGroupDurationSeconds) <
+			retentionRuleKey(string(obs.RetentionRules[j].Type), obs.RetentionRules[j].EverySeconds, obs.RetentionRules[j].ShardGroupDurationSeconds)
+	})
+	for i := range params.RetentionRules {
+		if params.RetentionRules[i].Type != string(obs.RetentionRules[i].Type) ||
+			params.RetentionRules[i].EverySeconds != obs.RetentionRules[i].EverySeconds ||
+			pointer.Int64Deref(params.RetentionRules[i].ShardGroupDurationSeconds, 604800) != pointer.Int64Deref(obs.RetentionRules[i].ShardGroupDurationSeconds, 0) {
```

Comment:
Why do we have different defaults here 🤔

Reply:
Given retentionRuleKey() returns a string incorporating all parameters, could we just build & compare it here?

Reply:
The problem is that this `ShardGroupDurationSeconds` has a default of `604800` but we are not really able to late-initialize since there is no unique identifier. I'll try to late-init with the key function and let you know.

Reply:
Alright, I used `Type` as the sole key for both up-to-date check and late-init. The documentation is really lacking but I believe this is unique across retention rules.